### DISCUSSION TITLE
Suggestions while typing

### DISCUSSION THREAD

**Discussion by teslum369** [cpeditor/cpeditor#822] [Ideas]
What do you think about suggestions while typing? Very often it's so handy and imo better way to type bcs it's not so typos prone. I thought about something like in sublime text 3 where we can use some suggestions if we used some keywords before.

Also i thought about something like shortcuts when typing - like when we type vi then vector<int> suggestion appears. 

Idk if that's possible to do / just hard to implement but i think that this feature would be very comfortable for programmers. What do you think about it?

**Comment by ouuan** [comment]
This has been requested in #156.

**Reply by teslum369** [reply]
Okay, can you close the suggestion section or remove it? I didn't Really notice it so now its like a spam. Sorry for that

**Reply by ouuan** [reply]
> Okay, can you close the suggestion section or remove it?

I don't get it. What's the "suggestion section"? However, this is fine and is definitely not spam. Actually, it helps people find this if they search "suggestions" instead of "autocomplete".

**Reply by ouuan** [reply]
If you meant this Discussion, unfortunately, it cannot be closed: https://github.com/github/feedback/discussions/3097.

**Reply by ouuan** [reply]
But Discussions are expected to be handled immediately. If it's worth further tracking, we'll open an issue for it.

**Reply by teslum369** [reply]
Okay, yes i meant this discussion

**Reply by teslum369** [reply]
Okay, so maybe one more idea / suggestion is to add something like corrector or autocomplection if we press key combination like ctrl + spacebar (or any other set).

Example:
cout << enld; 
after pressing ctrl + space (or anything other) it corrects to endl. Btw also if sb typed the variable name like: 'Index_of_f' after the shortcut it fills it to e.g. Index_of_first_array

I dont know if it was asked before tbh but i hope it will be a helpful suggestion.

**Reply by ouuan** [reply]
We'll add more LSP features in the future, including quick-fix, jump to def and so on. #156 was originally a request for LSP support but was later changed to a request for auto-complete by coder3101 when linting was added.

**Reply by coder3101** [reply]
> Okay, so maybe one more idea / suggestion is to add something like corrector or autocomplection if we press key combination like ctrl + spacebar (or any other set).
> 
> Example:
> cout << enld; 
> after pressing ctrl + space (or anything other) it corrects to endl. Btw also if sb typed the variable name like: 'Index_of_f' after the shortcut it fills it to e.g. Index_of_first_array
> 
> I dont know if it was asked before tbh but i hope it will be a helpful suggestion.

It is nothing different from automatically suggestion/completion as you type. In fact LSP servers `textdocument/completion` [request](https://microsoft.github.io/language-server-protocol/specification#textDocument_completion) asks for completion reason.

```typescript
export namespace CompletionTriggerKind {
	/**
	 * Completion was triggered by typing an identifier (24x7 code
	 * complete), manual invocation (e.g Ctrl+Space) or via API.
	 */
	export const Invoked: 1 = 1;

	/**
	 * Completion was triggered by a trigger character specified by
	 * the `triggerCharacters` properties of the
	 * `CompletionRegistrationOptions`.
	 */
	export const TriggerCharacter: 2 = 2;

	/**
	 * Completion was re-triggered as the current completion list is incomplete.
	 */
	export const TriggerForIncompleteCompletions: 3 = 3;
}
```

**Comment by coder3101** [comment]
Auto completion is already implemented from LSP client Library and it is able to send and get completion responses. Things that aren't working are because of QCodeEditor's bad/basic implementation of completion.

Here is the branch: https://github.com/cpeditor/cpeditor/tree/autocomplete-lsp

**Reply by rkscodes** [reply]
Is it working in recent beta versions or it's being worked on ?

**Reply by ouuan** [reply]
> Is it working in recent beta versions or it's being worked on ?

It's still being worked on.

**Reply by rkscodes** [reply]
ETA ?
